File: LeetCode/Medium/0103_Binary_Tree_Zigzag_Level_Order_Traversal/Solution.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 2022-05-21 (kinda hacky but w.e B))
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> order;
        zigzag(root, 0, order);
        for (int i = 1; i < order.size(); i += 2) 
            reverse(order[i].begin(), order[i].end()); // kinda hacky B)
        return order;
    }
    
    void zigzag(TreeNode* root, int level, vector<vector<int>>& order) {
        if (!root) return;
        
        if (order.size() <= level)
            order.push_back({});
        
        order[level].push_back(root->val);
        zigzag(root->left, level+1, order);
        zigzag(root->right, level+1, order);
    }
};

File: LeetCode/Medium/0103_Binary_Tree_Zigzag_Level_Order_Traversal/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

## 2023-02-18 (BFS)
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        input: a binary tree
        output: the zig-zag level order traversal of the nodes

        notes:
        - seems relatively straightforward? queue up level by level and process them
          in alternating directions
        - we can also start with an empty tree so b wary of that edge case
        - we can just repeatedly process the level as a stack, since that'd ensure we'd
          do a zig-zag pattern
            - actually we need to keep track of what direction we process as well since
              the order that we add the left/right children is also important
        
        analysis:
        - time = O(v) where v=number of nodes
        - space = O(v) where v=number of nodes
        """
        output = []
        stack = deque([root] if root else [])
        l_to_r = True
        while stack:
            output.append([])
            next_level = deque()
            while stack:
                node = stack[-1]
                stack.pop()
                order = (node.left, node.right) if l_to_r else (node.right, node.left)
                for child in order:
                    if child:
                        next_level.append(child)
                output[-1].append(node.val)
            stack = next_level
            l_to_r = not l_to_r
        return output
